app_dart/lib/src/request_handlers/refresh_chromebot_status.dart

```diff
@@ -79,19 +80,79 @@ class RefreshChromebotStatus extends ApiRequestHandler<Body> {
       DatastoreService datastore,
       Map<LuciBuilder, List<LuciTask>> luciTasks) async {
     for (String branch in branches) {
-      await for (FullTask task in datastore.queryRecentTasks(
-          taskName: builder.taskName, branch: branch)) {
-        for (LuciTask luciTask in luciTasks[builder]) {
+      final List<FullTask> tasks = await datastore
+          .queryRecentTasks(taskName: builder.taskName, branch: branch)
+          .toList();
+      final List<LuciTask> allLuciTasks = luciTasks[builder];
+      final Set<LuciTask> updatedLuciTasks = <LuciTask>{};
+
+      /// Scan both [tasks] and [luciTasks] from old to new, since [luciTasks] may
+      /// contain new re-run builds which are not in datastore yet. If matched, then
+      /// update accordingly.
+      for (int i = tasks.length - 1; i >= 0; i--) {
+        final FullTask task = tasks[i];
+        for (int j = allLuciTasks.length - 1; j >= 0; j--) {
```

Comment:
I guess these three loops are necessary. `branches` need to be iterated for sure. Then we need to find matched pairs of `luciTasks` and `tasks`. These two lists are ordered according to time by themselves, but no direct mapping between each other.

Reply:
Thanks for the `Map` suggestion, but it works when we have a strict mapping between a `datastoreTask` and a `luciTask`. 

However for a specific triple `{branch, commitSha, builder}`, there may be multiple runs (this is exactly why we create this PR to deal with the re-run cases). For such case, we have to iterate from the OLD to NEW to pair `datastoreTask` and `luciTask` one by one.

If we keep to use `Map`, it should be `{'branch': {'commitSha': {'builderName', ['buildList']}}}`, but it doesn't help reduce the loop cycles as I see it.

Due to the above mentioned `from OLD to NEW` iteration requirement, we have to iterate the ordered `datastoreTasks` list queried according to a `branch`-`builder` pair, rather than randomly iterate through. Based on the returned `datastoreTasks` list, we then try finding the paired `luciTask`.

Reply:
Using a map was a suggestion not the final solution. 4 nested loops strongly suggests problems in the way we are getting the data or the way we are processing the data I'd suggest to model classes, queries to buildbucket or queries to datastore to simplify this.

Reply:
Managed to reduce one loop level by changing the returned `luciTasks` to consider `branch`. With this, we need to iterate only `branch` list occurred in these `luciTasks`, rather than iterating all `branch` options. Thank you for the suggestion.